### ISSUE TITLE
"Avoid using undefined variables" for array initialization with offset.

### ISSUE THREAD

**Issue by donquixote** [phpmd/phpmd#1077]
- PHPMD version: 2.15.0
- PHP Version: 8.1.23
- Installation type: composer
- Operating System / Distribution & Version: Docker with some kind of Linux

### Current Behavior
phpmd reports "VIOLATION": " Avoid using undefined variables such as '$form' which will lead to PHP notices."

### Expected Behavior
No complaints from phpmd.

### Steps To Reproduce:
Code like this:
```php
function f(): array {
  $form['x'] = 'y';
  return $form;
}
```

### Notes
Initializing an array like this is valid php, and is common in e.g. the Drupal ecosystem.
We can call `$form = [];` before, but it would be redundant.

At least this case should be optional, and use a different message, because saying "which will lead to PHP notices." is wrong here.

## Checks before submitting
* [x] Be sure that there isn't already an issue about this. See: [Issues list](https://github.com/phpmd/phpmd/issues)
  -> I searched for "Avoid using undefined variables array" and did not find an issue that seemed to match the problem.
* [x] Be sure that there isn't already a pull request about this. See: [Pull requests](https://github.com/phpmd/phpmd/pulls)
  -> nope
* [x] I have added every step to reproduce the bug.
* [x] If possible I added relevant code examples.
* [x] This issue is about 1 bug and nothing more.
* [x] The issue has a descriptive title. For example:  "JSON rendering failed on Windows for filenames with space".

**Comment by kylekatarnls** [comment]
This is expected behavior, you're mutating an array that does not exist and PHP being resilient create an empty array for you, but it's generally considered best-practice to actually create the empty array explictly:

```php
function f(): array {
  $form = [];
  $form['x'] = 'y';
  return $form;
}
```

I guess the code above is a simplified example but of course in this precise case it's even better to do it in a single operation by creating the array full shape:
```php
function f(): array {
  $form = ['x' => 'y'];
  return $form;
}
```
And once there the intermediate variable becomes useless:
```php
function f(): array {
  return ['x' => 'y'];
}
```

**Comment by donquixote** [comment]
> it's generally considered best-practice to actually create the empty array explictly:

How can we verify if something is "generally considered best practice"?
I think it is more a matter of convention.

Even if it is "best practice", the part saying "which will lead to PHP notices" is clearly not true in this case.
More accurate would be "It is best practice to initialize arrays before assigning an array value.". And it could be a separate rule "ArrayNotInitialized" that can be disabled separately.
(In practice it is fine to disable the complete "UndefinedVariable" rule, if there are other tools like phpstan that can detect undefined variables.)

> `$form = ['x' => 'y'];`

Sure this works, but there can be many reasons to prefer `$form['x']['y']['z'] = 'XYZ';`. Especially if we have a sequence of similar assignments.

I found [this reddit discussion](https://www.reddit.com/r/PHP/comments/br5gfg/do_you_need_to_declare_an_array_before_you_start/).
There is an argument that we are not sure if the array already contained some values to start with.
But this does not really apply to short functions, where we can see that a variable did not exist before.

Ofc it is different if we have a loop, or all the assignments are conditional, in that case we need to initialize the empty array first, to make sure it is defined.

**Comment by AJenbo** [comment]
> How can we verify if something is "generally considered best practice"?

Experience, survey various tools, guides, and developers.

> I think it is more a matter of convention.

Well that is kinda what is implied by "generally".

```php
// Maybe this is a bug where the array was expected to already exist?
$form['x']['y']['z'] = 'XYZ';
// Clearly creating a new array, no bug here
$form = ['x' => ['y' => ['z' => 'XYZ']]];

// Repeating the same code can easily lead to hard to spot errors
$form['x']['y']['a'] = 'XTZ';
$form['x']['y']['b'] = 'XVZ';
$form['x']['u']['c'] = 'XIZ';
$form['x']['y']['d'] = 'XLZ';
$form['x']['y']['e'] = 'XAZ';

// No bugs in this one, also less code
$form = ['x' => ['y' => [
    'a' => 'XTZ',
    'b' => 'XVZ',
    'c' => 'XIZ',
    'd' => 'XLZ',
    'e' => 'XAZ',
]]];
```

> Even if it is "best practice", the part saying "which will lead to PHP notices" is clearly not true in this case.

True, maybe it's a general notice for this rule? Maybe a better rephrasing would be "will" -> "can".

>  (In practice it is fine to disable the complete "UndefinedVariable" rule, if there are other tools like phpstan that can detect undefined variables.)
> ...
> There is an argument that we are not sure if the array already contained some values to start with.

Sounds like you are already aware that this is generally considered best practice.

> But this does not really apply to short functions, where we can see that a variable did not exist before.

What is the general rule for what defines a short function where that can be deduced. Are you proposing that it be implemented as an exception, and why have inconsistent ways of writing code depending on if it's in a short or long function when it does the same. You would have to change it if you copied it or the function grew.

> Sure this works, but there can be many reasons to prefer $form['x']['y']['z'] = 'XYZ';. Especially if we have a sequence of similar assignments.

I don't think this is true, see my example.

**Comment by AJenbo** [comment]
The examples given by kyle also show that it generates less op code:
https://3v4l.org/6P0JU/vld
vs
https://3v4l.org/3TNKD/vld
